docs/architecture/architecture.md

```diff
@@ -0,0 +1,674 @@
+
+On the host, each container process's removal is handled by a reaper in the higher
+layers of the container stack. In the case of Docker it is handled by `containerd-shim`.
+In the case of CRI-O it is handled by `conmon`. For clarity, for the remainder
+of this document the term "container process reaper" will be used to refer to
+either reaper. As Clear Containers processes run inside their own  virtual machines,
+the container process reaper can not monitor, control
+or reap them. `cc-runtime` fixes that issue by creating an [additional shim process
+(`cc-shim`)](https://github.com/clearcontainers/shim) between the container process
+reaper and `cc-proxy`. A `cc-shim` instance will both forward signals and `stdin`
+streams to the container process on the guest and pass the container `stdout`
+and `stderr` streams back up the stack to CRI-O or Docker via the container process
+reaper. `cc-runtime` creates a `cc-shim` daemon for each container and for each
+OCI command received to run within an already running container (example, `docker
+exec`).
+
+The container workload, that is, the actual OCI bundle rootfs, is exported from the
+host to the virtual machine.  In the case where a block-based graph driver is
+configured, virtio-blk will be used. In all other cases a 9pfs virtio mount point
+will be used. `cc-agent` uses this mount point as the root filesystem for the
+container processes.
+
+![Overall architecture](arch-images/overall-architecture.png)
+
+## Hypervisor
+
+Clear Containers use  [QEMU](http://www.qemu-project.org/)/[KVM](http://www.linux-kvm.org/page/Main_Page)
+to create virtual machines where containers will run:
+
+![QEMU/KVM](arch-images/qemu.png)
+
+Although Clear Containers can run with any recent QEMU release, Clear Containers
+boot time and memory footprint are significantly optimized by using a specific QEMU
+version called [`qemu-lite`](https://github.com/01org/qemu-lite/tree/qemu-2.7-lite).
+
+Clear Containers supports various machine types, including `pc`, `pc-lite` and `q35`.
+Clear Containers defaults to using the `pc` machine type. In the past
+`pc-lite` was utilized which provided the following improvements:
+- Removed many of the legacy hardware devices support so that the guest kernel
+  does not waste time initializing devices of no use for containers.
+- Skipped the guest BIOS/firmware and jumped straight to the Clear Containers
+  kernel.
+
+In order to provide some advanced features which were lacking in `pc-lite` like hotplug
+support, the default was changed to `pc`.
+
+In the future, Clear Containers plan to move to a 2.9 based version of QEMU,
+available at 
+the [Clear Containers QEMU repo](https://github.com/clearcontainers/qemu/tree/qemu-lite-v2.9.0).
+This transition has been delayed until after the release of Clear Containers 3.0
+due to regressions, as described in [runtime issue 407]
+(https://github.com/clearcontainers/runtime/issues/407). Once support for
```

Comment:
Breaking up this [runtime issue 407](https://github.com/clearcontainers/runtime/issues/407) breaks the link's markup. Don't worry about going over 78chars when it's code or markup forcing you over:
due to regressions, as described in [runtime issue 407](https://github.com/clearcontainers/runtime/issues/407).
Once support for

Reply:
I meant to fix all those -- thanks!  Done